Utility functions for option parsing, validity checks, and pretty printing errors/warnings

faithokamoto opened the pull request:
## Changelog Entry
To be copied to the [draft changelog](https://github.com/vgteam/vg/wiki/Draft-Changelog) by merger:

 * Create utility functions for basic parsing/validity checks, and use them in subcommands + `src/index_registry.cpp`
     * `parse_thread_count()`, `parse_split_string()`, and `assign_fastq_files()` are meant to accept `optarg` and then do common nontrivial parsing
     * `error_if_file_does_not_exist()`, `error_if_file_cannot_be_written()`, and `error_if_file_is_gzipped()` are meant to accept `optarg` and perform standardized checks for file validity
* At least attempt to enforce the use of some new standardized parsing/validity functions
* Create utility functions for pretty error/warning printing, and use them in subcommands + `src/index_registry.cpp`
* Fix two broken tests in `test/t/03_vg_view.t`

## Description

Welcome back to Faith's Nitpick Corner, where today we are serving a sizzling side project of centralized utility functions.

### Motivation

Many `vg` subcommand options have similar basic parsing logic. In the best-case scenario, this led to copious copy-pasting, which is why so many subcommands have e.g. [the](https://github.com/vgteam/vg/blob/e78e9d57dbd9777c9430ae0eb21ca6f82b880b4f/src/subcommand/paths_main.cpp#L306) [same](https://github.com/vgteam/vg/blob/e78e9d57dbd9777c9430ae0eb21ca6f82b880b4f/src/subcommand/zipcode_main.cpp#L127) [code](https://github.com/vgteam/vg/blob/e78e9d57dbd9777c9430ae0eb21ca6f82b880b4f/src/subcommand/call_main.cpp#L303) for parsing `--threads`. In the worst-case scenario, people didn't bother adding validity checks. In the confusing-case scenario, people copy-pasted badly, which is how you get [`vg augment` claiming to be `vg call`](https://github.com/vgteam/vg/blob/e78e9d57dbd9777c9430ae0eb21ca6f82b880b4f/src/subcommand/augment_main.cpp#L233).

Initially this was only supposed to be `parse_thread_count()` (errors if given nonpositive input) and `error_if_file_does_not_exist()` (errors if file can't be read). Then it ballooned into adding functions that check for if a file is gzipped (`error_if_file_is_gzipped()`), use the same option to accept up to two FASTQ files (`assign_fastq_files()`), accept an argument of the form `X<delimiter>Y` and extract `X` and `Y` (`parse_split_string()`), or check if a file can be _output_ to (`error_if_file_cannot_be_written()`).

Since all of these new utility functions needed to print out errors, I also made `error_and_exit()` and `emit_warning()` to print an error/warning in a standardized format via `emit_with_indent()`. `error_and_exit()` additionally exits with an error code. These pretty printing functions I used all over the subcommands as well as `src/index_registry.cpp`. I don't _have_ to use them; I just like them. I could revert to the previous error/exit stuff if we want. Though if we do that, we should make sure that all errors actually *error* ([example](https://github.com/vgteam/vg/blob/e78e9d57dbd9777c9430ae0eb21ca6f82b880b4f/src/subcommand/gbwt_main.cpp#L682)).

### Fiddly bits

* Currently `error_if_file_cannot_be_written()` checks if a file can be written to by attempting to open a file for writing. This means it will overwrite whatever was there before. I had to [change a test](https://github.com/vgteam/vg/commit/7f4171cd8f99e05eaaf325ce0a1f2e64db2bb3d1) to move an error test to later, to prevent it from overwriting a file used by a different test.
* I belatedly realized that `split_delims()` is a thing that [already exists](https://github.com/vgteam/vg/blob/e78e9d57dbd9777c9430ae0eb21ca6f82b880b4f/src/utility.cpp#L191-L209), but I think `parse_split_string()` is still useful for its simpler interface in an argument parsing context.
* In places where I'd already validated a filename validity during argument parsing, I removed later duplicate validity checks ([example](https://github.com/vgteam/vg/blob/e78e9d57dbd9777c9430ae0eb21ca6f82b880b4f/src/subcommand/add_main.cpp#L159)). I think it's better to have the validity checks centralized and, crucially, _early_: this way we'll error before doing any expensive algorithms.
* While I _generally_ kept the error/warning messages the same as before, some of the `vg mpmap` error messages used the wrong shortforms for options ([example](https://github.com/vgteam/vg/blob/e78e9d57dbd9777c9430ae0eb21ca6f82b880b4f/src/subcommand/mpmap_main.cpp#L1249)), which I fixed.
* The pretty printing functions are [thread-safe](https://github.com/vgteam/vg/blob/27a9042e8c53d36c0f72a3f7784bedae58c656a4/src/utility.cpp#L28), which feels slightly overkill but at least one of the subcommand files had a thread-safe error print so I ported that functionality.
* I had to slightly rewrite some tests to make them expect my new pretty-printed errors. Their intent/purpose was not changed.
* [Two tests](https://github.com/vgteam/vg/blob/e78e9d57dbd9777c9430ae0eb21ca6f82b880b4f/test/t/03_vg_view.t#L65-L73) in `test/t/03_vg_view.t` were accessing bad file names, which led to empty output. Since these empty outputs were compared to each other, the test passed. I swapped to the correct file names.
* I used `error_if_file_does_not_exist()` in `IndexRegistry`. To then get it past unit tests which intentionally use dummy files, I added a flag to the object which tells it to skip checking the files passed to `provide()`.
* `scripts/check_options.py` attempts to enforce the usage of `error_if_file_does_not_exist()`, `error_if_file_cannot_be_written()`, and `parse_thread_count()`. Unfortunately it has to use heuristics for all of them: filenames variables are auto-detected as `<variable name ending> = optarg` where the known endings are in `FILENAME_VAR_ENDS`. This will fail to catch other name formats such as `xg_name = optarg` as `_name` would lead to false positives. It also won't catch things like `vcf_filenames.append(optarg)` where there's a list of filenames in use. The thread count heuristic requires `omp_set_num_threads()` and `parse_thread_count()` to be on the same line.

adamnovak commented on src/utility.hpp:
```diff
@@ -25,6 +25,21 @@ namespace vg {
+/// Warn the user with a standard format
+void emit_warning(const string& context, const string& message);
+
+/// Write a message to an output stream with indentation
+/// Inserts "indent" spaces on the beginning of each line
+/// Honors existing newlines, but otherwise adds newlines between words
+/// such that everything gets printed with width no more than 80
+/// Assumes that the first line is already indented
+/// (because you put something there and you want others to line up)
+void emit_with_indent(ostream& outstream, const string& message, size_t indent);
```
This doesn't promise that things like dumped JSON, FASTA sequence headers, malformed GFA lines, or weird but revealing garbage where the program expected important data to be, will survive the reflow process intact or copy-paste-able. If we feed in something like:
```
warning: we couldn't handle quality string because it is malformed: character ' ' is not allowed in the string.
```
I think that would wrap to the somewhat confusing:
```
warning: we couldn't handle quality string because it is malformed: character '
         ' is not allowed in the string.
```
Which is a lot harder to grasp.

And JSON with spaces in quoted strings could wrap to JSON that can't actually be loaded, so `error_and_exit(context, "There's a very big problem with this alignment: " + pb2json(alignment))` would potentially get into trouble if somehow there were strings with spaces in there.

*Mostly* things would work, though, and maybe we shouldn't be dumping anything we need to be machine-readable through the system we're using for human-readable errors?

We also don't know that the user's terminal is 80 characters wide, but we do have access to APIs that will tell us the current width of the terminal, or whether our output stream is connected to the terminal at all. If the user has a narrower terminal, we probably don't want to be wrapping to 80 characters, because that will be harder to read than just letting the terminal reflow the text.

We can't really do anything with long nucleotide sequences that might appear and upset the formatting. (But maybe we could if we were *really* fancy and sniffed for them?) And if we output an actual tab character, the extra width (4 or 8 characters?) isn't accounted for in the wrapping.

adamnovak commented on src/utility.hpp:
```diff
@@ -821,6 +835,40 @@ string file_base_name(const string& filename);
+/// Will actually open the file, so it will overwrite the current contents.
+bool file_can_be_written(const string& filename);
+
+/// Check if a file exists and return its name (if so) or error.
+/// Uses file_exists() and is intended to be called when parsing arguments.
+string error_if_file_does_not_exist(const string& context, const string& filename);
```
This function and the others that have it here should probably explain their `context` argument. I guess it's the context to attach to any errors?

It might also be good to specify that this will "print an error message and stop the program", rather than `throw`ing an error, because just saying it will "return ... or error" makes it sound like it might throw or like it might even return an error message.

faithokamoto replied:
implemented this

adamnovak commented on src/utility.hpp:
```diff
@@ -821,6 +835,40 @@ string file_base_name(const string& filename);
+string error_if_file_does_not_exist(const string& context, const string& filename);
+
+/// Error if a file looks like it's gzipped.
+void error_if_file_is_gzipped(const string& context, const string& filename);
```
Why doesn't this one return the file's name like the others do?

faithokamoto replied:
mostly because it's usually called after `error_if_file_does_not_exist()` but I'll change it for consistency

adamnovak commented on src/utility.hpp:
```diff
@@ -821,6 +835,40 @@ string file_base_name(const string& filename);
+/// Else if the second file is unset, then it gets the input filename
+/// If both are already set, then this errors
+/// Also calls error_if_file_does_not_exist() on the filenames
+void assign_fastq_files(const string& context, const string& input_filename, string& fastq1, string& fastq2);
```
OK, so a proper usage of this function involves calling it twice? Maybe it should have a name that reflects that it only does a single file assignment somehow?

I might write this:
```suggestion
void push_fastq(const string& context, const string& input_filename, string& slot1, string& slot2);
```
Or
```suggestion
void add_fastq(const string& context, const string& input_filename, const std::pair<std::string, std::string>& fastq_pair);
```

adamnovak commented on src/utility.hpp:
```diff
@@ -821,6 +835,40 @@ string file_base_name(const string& filename);
+
+/// Parse an argument that should be a pair of strings separated by a delimiter.
+/// Errors if the string does not contain exactly one delimiter.
+pair<string, string> parse_split_string(const string& context, const string& arg,
+                                        const char& delimiter, const string& option_name);
```
If this is a pair-only function and not a general split, maybe it should be `parse_pair`?

Since it gets [used to parse things that aren't *really* strings and immediately get `parse<>`'d themselves](https://github.com/vgteam/vg/pull/4689/files#diff-732a7706426698185d84aac919c3b444e8748a107ee89a367645e937d24f47ddR177-R180), this should maybe integrate with that system and become more like:

```suggestion
template<typename T1 = std::string, typename T2 = std::string>
std::pair<T1, T2> parse_pair(const string& context, const string& arg, const char& delimiter, const string& option_name);
```

Then it can call `parse<>` to build the pair items internally (maybe with a `parse<string>` specialization getting defined that does nothing), and you can `parse_pair<string, size_t>(...)` when that's what you need, and it will Just Work.

adamnovak commented on src/utility.hpp:
```diff
@@ -821,6 +835,40 @@ string file_base_name(const string& filename);
+/// Determine if a file can be written to.
+/// Only works for files writable by the current user.
+/// Meant to be used to validate output file names.
+/// Will actually open the file, so it will overwrite the current contents.
+bool file_can_be_written(const string& filename);
```
I'm worried about having this function blow away the current contents of the file as a side effect of the test.

It's hard for me to come up with a good example of how it would go wrong, though; the best I have is if you're going to run a command that would regenerate several files, and one of them isn't writable, so it aborts as soon as it notices, but at that point it has already touched and clobbered the other files and you have neither your old nor your new index. Or maybe in a similar case where your command line is very wrong and you've put an input file in an output slot, and the command figures out that you've asked for obvious nonsense, but not before blowing away the input file.

It makes it hard to use this in something like `vg add-to-database whatever.db input.dat`, but we don't actually use that pattern at all.

It's not *worse* than `>`'s behavior, but I feel like it could be better. *Actually* opening and writing to the file, and then closing it again and returning `true`, doesn't help at all with time-of-check/time-of-use confusion, because we're just going to re-open the file again by name later.

adamnovak commented on src/utility.hpp:
```diff
@@ -821,6 +835,40 @@ string file_base_name(const string& filename);
+/// A special parser for thread count which errors if non-positive
+/// If max_threads is non-zero, it will also decrease to that maximum.
+/// If max_threads is zero, max is omp_get_max_threads()
+int parse_thread_count(const string& context, const string& arg, int max_threads = 0);
```
Rather than enforcing that `omp_set_num_threads()` and `parse_thread_count()` are used together with a checker script, why not just make one function that sets the thread count from a string and always call that?

adamnovak commented on src/utility.hpp:
```diff
@@ -821,6 +835,40 @@ string file_base_name(const string& filename);
+/// Uses file_exists() and is intended to be called when parsing arguments.
+string error_if_file_does_not_exist(const string& context, const string& filename);
+
+/// Error if a file looks like it's gzipped.
```
This should maybe specify that we're looking at the name, and we're not sniffing for the GZip magic numbers/headers, so this is safe to use on a `.bam` but will maybe not do what is expected on a `.tgz`.

adamnovak commented on src/utility.hpp:
```diff
@@ -821,6 +835,40 @@ string file_base_name(const string& filename);
 /// Only works for files readable by the current user.
 bool file_exists(const string& filename);
 
+/// Determine if a file can be written to.
+/// Only works for files writable by the current user.
+/// Meant to be used to validate output file names.
+/// Will actually open the file, so it will overwrite the current contents.
+bool file_can_be_written(const string& filename);
+
+/// Check if a file exists and return its name (if so) or error.
+/// Uses file_exists() and is intended to be called when parsing arguments.
+string error_if_file_does_not_exist(const string& context, const string& filename);
+
+/// Error if a file looks like it's gzipped.
+void error_if_file_is_gzipped(const string& context, const string& filename);
+
+/// Check if a file can be written to and return its name (if so) or error.
+/// Uses file_can_be_written() and is intended to be called when parsing arguments.
+string error_if_file_cannot_be_written(const string& context, const string& filename);
```
How do these various file-checking functions interact with "-", which we can use in many places as representing "read from standard input"? (For example, I'm pretty sure you can `vg filter --exact-name -n MyRead reads.gam | vg view -dp graph.vg -A - | dot -Tsvg -o graph_with_read.svg` in order to have `vg view` read a graph from a file and a read to visualize on it from standard input.

Probably we want "-" to *not* appear writable, even though you probably could create and write to `./-`. And we would want it to appear as if it *does* exist, because if we use any of the `get_input_file()` functions to open it, it effectively exists.

But then, are the checkers always used with those accessors that support "-"?

adamnovak commented on src/utility.hpp:
```diff
@@ -821,6 +835,40 @@ string file_base_name(const string& filename);
+
+/// Check if a file exists and return its name (if so) or error.
+/// Uses file_exists() and is intended to be called when parsing arguments.
+string error_if_file_does_not_exist(const string& context, const string& filename);
```
The names of these `error_if_...` functions don't really suggest that you can use them as type-narrowing filters in the middle of expressions (i.e. the argument is a file that may not exist, but the return type is a file constrained to always exist), but that seems to be the way they're mostly used, and probably the best way to use them. Also, the names are a bit long for something we might use in more or less every flag that takes a filename.

I would call these something more like`require_exists()`, `require_writable()`, etc. Not `ensure`, since that sounds like it would make a file exist when it didn't, but also not a whole clause for something that's meant to be used so frequently, if it can be avoided, and ideally something that sounds like part of an expression and not a code block.

faithokamoto replied:
I'm swapping to `require_exists()` and `ensure_writable()`. The latter is "ensure" because it does actually make a file exist if it didn't exist before.

adamnovak commented on src/utility.cpp:
```diff
@@ -24,6 +24,67 @@
 
 namespace vg {
 
+void error_and_exit(const string& context, const string& message) {
```
This might be better as a `fatal_error()`?

faithokamoto replied:
renamed

adamnovak commented on src/utility.cpp:
```diff
@@ -24,6 +24,67 @@
+    exit(EXIT_FAILURE);
+}
+
+void emit_warning(const string& context, const string& message) {
```
Would `warning()` or `warn()` alone be too dangerous to have kicking around our namespace?

faithokamoto replied:
I named it `warning()`, seems fine

adamnovak commented on src/subcommand/sift_main.cpp:
```diff
@@ -557,33 +556,27 @@ int main_sift(int argc, char** argv){
-        if (is_paired){
-            cerr << "Processing..." << endl;
+        if (is_paired) {
+            cerr << context << ": Processing..." << endl;
```
Here this is touching essentially info-level logging lines, which is making it smell more and more like it's actually a logging framework.

adamnovak reviewed (changes requested):
This is a great idea!

I have a bunch of design comments and quibbles with the names of things.

The biggest issue is that basically now you've written 70-80% of a logging framework. Which maybe should go all the way to 100% of our own logging framework (`log.hpp`, documentation for "How logging should work in vg" and "why we need our own logging framework", etc.), or maybe should go to 0% and have us just import a real logging framework someone else maintains that meets our needs.

faithokamoto commented on src/utility.hpp:
```diff
@@ -25,6 +25,21 @@ namespace vg {
 
 using namespace std;
 
+/// Error (and crash) with a standard format
+/// If you're erroring because a file doesn't exist,
+/// consider using error_if_file_does_not_exist() instead.
+void error_and_exit(const string& context, const string& message);
+/// Warn the user with a standard format
+void emit_warning(const string& context, const string& message);
```
I couldn't find a library that did something like this - logging frameworks tended to be way too fancy and aren't meant to support the simple one-line streams that we want for these functions. So I rolled my own. (Possibly I didn't look hard enough because I had an idea for rolling my own and I liked it... sorry.)

Glad to say that
```
warning(context) << "something is bad" << std::endl;
fatal_error(context) << "something is bad and we'll somehow work out when to kill the program because of it" << std::endl;
```
both now "just work". In actuality those functions will do a set-up log to `cerr` and then return a `cerrWrapper` that passes anything `<<`-ed to `cerr` itself. For `fatal_error()`, once the `cerrWrapper` goes out-of-scope, its destructor handles the clean program exit. For when we want to build up a longer message, we can save the initial `cerrWrapper` object and then stream to it as if it were `cerr`. 

I _wanted_ to make indenting work, but it got annoying and I decided to leave that For Later (TM).  Some things that I found include [a library that handles indenting](https://github.com/asap-projects/asap-textwrap) (but seems to stick everything in an in-between string?) and that, if we figure out how to make it work, I could stick a `stringbuf` inside the `cerrWrapper` to [handle the indenting](https://stackoverflow.com/a/9600752/). The difficult part is figuring out when to do the indents once you move to a paradigm where each char is passed to `cerr` one by one.

faithokamoto replied:
(passing everything inside of `cerrWrapper` is meant to let us change it later easily)

faithokamoto replied:
though possibly it makes more sense to return a plain `cerr` for the ones that don't need to kill the program during self-destruct

faithokamoto commented:
Something else I recall, as I slowly double-check all my file changes, is that I replaced some [runtime *errors*](https://github.com/vgteam/vg/blob/dfe612c34248a2b6615cd9ee84b298fb2e887b82/src/subcommand/msga_main.cpp#L616-L619) (as in the things you raise/[yeet](https://github.com/vgteam/vg/pull/2601)) with [vanilla program exits](https://github.com/vgteam/vg/blob/203906ecd10a516f206308165501dff41c31e155/src/subcommand/msga_main.cpp#L617-L619). Should those be changed back?

faithokamoto commented:
Hey, just over a month later, I think I've addressed your issues @adamnovak 
* Stuff got renamed (`require_exists()`, `ensure_writable()`, `parse_pair()` etc.) and the interfaces jiggered to make them easier to use (`parse_pair()` now uses templates, the logging functions have `<<` streaming that Just Works, `set_thread_count()` parses and sets the thread count, etc.)
* `ensure_writable()` doesn't overwrite files any more - if it detects that a file already exists, it just checks permissions
* No pretty printing because I couldn't figure out a good way to do it. Back to everything being on one line by default. (I considered modifying the magic streaming interface do auto-indent whenever the user streamed a newline, but not sure if that would be wanted.)
* Logging things moved to `log.cpp`/`log.hpp`
Basically, I think I addressed all the last round of comments. No idea why the Mac test failed though :(

adamnovak commented:
It looks like the mac tests immediately failed linting, but then went on to do a bunch of other stuff because linting and building are in parallel.

```
 src/subcommand/chains_main.cpp: Use require_exists() or ensure_writable() for standardized file checks: this->output_file = optarg;
make: *** [lint] Error 1
make: *** Waiting for unfinished jobs....
```

adamnovak reviewed (changes requested):
I think I found one place where the log message won't be quite right.

I think the logging framework makes sense. I'm not sure about naming what are essentially the 3 log levels `basic_log`, `warning`, and `fatal_error`, since those aren't clearly a set. And if there were a good way to get out of always passing the context that would be nice, but I'm not sure there is.

(Python as used in Toil has you set up a global `logger` object at the top of every file, with the file's name as context, and then you call `logger.info()` and `logger.warning()` and so on to log at different levels. But in vg we like to also include the function name. C++ has that in `__func__` but you would need some kind of macro to get at it and it wouldn't always be appropriate.)

faithokamoto commented on src/log.hpp:
```diff
@@ -0,0 +1,79 @@
+        return *this;
+    }
+
+    cerrWrapper& operator<<(std::ostream& (*manip)(std::ostream&)) {
```
We need that to handle `std::endl`.

adamnovak replied:
Well yes, but the question the code raised for me is why `std::endl` won't just have some (function pointer) type that the first `const T& t` template can be instantiated on.

I think the answer is that `std::endl` is itself a template and needs to use template type inference to pick up its template parameters from the argument type of the `operator<<` it is passed to. If the operator and `std::endl` both try to infer off each other, nobody knows what the stream character type is.

We could do something like having an `operator<<` that takes a *template* template parameter, alongside the one that takes a type, but if we have to have two copies anyway we might as well just make a concrete overload for the one thing we actually need.

faithokamoto commented on src/log.hpp:
```diff
@@ -0,0 +1,79 @@
+        this->str = &std::cerr;
+    }
+
+    template <typename T>
+    cerrWrapper& operator<<(const T& t) {
+        if (str) {
+            *str << t;
+        }
+        return *this;
+    }
+
+    cerrWrapper& operator<<(std::ostream& (*manip)(std::ostream&)) {
+        *str << manip;
+        return *this;
+    }
```
good point that was a blind copy-paste issue. changed.

faithokamoto commented on src/log.hpp:
```diff
@@ -0,0 +1,79 @@
+// https://stackoverflow.com/a/25615354/
+class cerrWrapper {
+private:
+    std::ostream* str;
```
well I got rid of the variable anyhow.

faithokamoto commented on src/log.hpp:
```diff
@@ -0,0 +1,79 @@
+
+/// Log to cerr with a standard format
+/// "context" is caller context, e.g. "[vg inject]"
+cerrWrapper basic_log(const std::string& context);
```
`print()` works - what do you think about that?

adamnovak replied:
I think `print(const std::string&)` is a bad idea because that would look weird when you use it as a stream.

adamnovak replied:
I think the C++ approach would be to put all this stuff in a `vg::log` namespace and then `using vg::log;` to make it shorter when we want.

If I designed it I think I would have a `Logger` class you construct with the `const std::string& context`, and then in each file or in each function I would have `Logger log(...);` and then say `log.info() << ...`.

Or maybe I would have macros for the front end to capture the function name magically and I would say `INFO << ...`.

If we just want to name these things, though, I would go with `log_info`, `log_warning`, and `log_fatal_error`, so they all match but you can still tell fatal errors are magically fatal.

faithokamoto commented:
Agh and right after I'd merged master into my branch, something else gets merged into master and causes conflicts. I'll wait until you say what other changes might be needed before resolving, Adam.

adamnovak commented on src/index_registry.cpp:
```diff
@@ -319,7 +314,7 @@ vector<string> vcf_contigs(const string& filename) {
     htsFile* vcf = hts_open(filename.c_str(),"rb");
     if (vcf == nullptr) {
-        cerr << "error:[IndexRegistry] Could not open VCF" << filename << endl;
+        error(context) << "Could not open VCF: " << filename << endl;
```
I guess having `error()` exit helps us avoid cases like we're fixing here, where an error is detected but not acted on.

But now that this is so simple and in some ways so much like Python logging it's a little astonishing that that happens. Maybe we just have to see if it actually trips people up ever?

adamnovak reviewed (changes requested):
I think I found one actual wrong line, and a couple things where there might not be anything to do.

faithokamoto commented:
And I have no clue how that non-standard `--help` description managed to sneak past `check_options.py` to be merged in.
